docs/adr/0006_component_constructors.md

```diff
@@ -0,0 +1,334 @@
+}
+
+
+// SourceInputMethod is the interface for processing a source with an input method declared as per
+// [spec.Source.Input]. Note that spec.Source's who have their access predefined, are never processed
+// with a ResourceInputMethod, but are directly added to the component version repository.
+// any spec.Source passed MUST have its [spec.Source.Input] field set.
+// If the input method does not support the given input specification it MAY reject the request
+//
+// The method will get called with the raw specification specified in the constructor and is expected
+// to return a ResourceMethodResult or an error.
+//
+// A method can be supplied with credentials from any credentials system by requesting a consumer identity with
+// SourceInputMethod.GetCredentialConsumerIdentity. The resulting identity MAY be used to uniquely identify the consuming
+// method and to request credentials from the credentials system. The credentials system is not part of this interface and
+// is expected to be supplied by the caller of the input method.
+//
+// The resolved credentials MAY be passed to the input method via the credentials map, but a method MAY
+// work without credentials as well.
+type SourceInputMethod interface {
+    GetCredentialConsumerIdentity(ctx context.Context, source *constructor.Source) (identity runtime.Identity, err error)
+    ProcessSource(ctx context.Context, source *constructor.Source, credentials map[string]string) (result *SourceInputMethodResult, err error)
+}
+```
+
+The contract explicitly opens the possibility to integrate our existing systems such as our credential graph.
+However, because methods must be available at runtime to the constructor library, we also support a set of provider
+interfaces that can be integrated with our plugin system.
+
+```go
+type ResourceInputMethodProvider interface {
+    // GetResourceInputMethod returns the input method for the given resource constructor specification.
+    GetResourceInputMethod(ctx context.Context, resource *constructor.Resource) (input.ResourceInputMethod, error)
+}
+
+type SourceInputMethodProvider interface {
+    // GetSourceInputMethod returns the input method for the given source constructor specification.
+    GetSourceInputMethod(ctx context.Context, resource *constructor.Resource) (input.ResourceInputMethod, error)
+}
+```
+
+To allow for actual construction of component versions, the constructor library provides a central entry function:
+
+```go
+// TargetRepository defines the interface for a target repository that can store component versions created 
+// by the constructor but also act as a target for the constructor to add resources to.
+type TargetRepository interface {
+    // AddLocalResource adds a local resource to the repository.
+    // The resource must be referenced in the component descriptor.
+    // Resources for non-existent component versions may be stored but may be removed during garbage collection.
```

Comment:
How does it know that the component version doesn't have this resource? And how do you even add a Resource without a component version? I would think that having a component version to add a resource is sort of like a prerequisite?

Reply:
Resources are independent of component version lifecycles. Local Resources have to be added BEFORE a component version. If a blob gets added twice, then the call should be indempotent

Reply:
Ah, yah, that makes sense now.